sql_insert_writer/sql_insert_writer.py

```diff
@@ -0,0 +1,241 @@
+    return '\n'.join(from_clause)
+
+
+def cast(column_str, new_type, db_url):
+    'Produces a version of `column_str` cast to data type `new_type` in `db_url` dialtect'
+    engine_name = db_engine_name(db_url)
+    if engine_name == 'postgresql':
+        return '{}::{}'.format(column_str, new_type)
+    elif engine_name == 'mysql':
+        return 'CAST({} AS {})'.format(column_str,
+                                       new_type)  # TODO: include data length
+
+
+def generate_from_tables(db_url,
+                         destination,
+                         sources,
+                         qualify=False,
+                         type_cast=False):
+    """
+    Generates an `INSERT INTO... SELECT FROM` SQL statement.
+
+    Args:
+        db_url (str): Database URL in SQLAlchemy format
+        destination (str): Name of table to INSERT into
+        sources (list): Names of tables to select from, in order of preference
+        qualify (bool): Qualify column names with table name even if only one table
+        type_cast (bool): Cast values to destination data type where needed
+
+    Returns:
+        str: A SQL statement
+
+    """
+
+    db = records.Database(db_url)
+
+    dest_column_block = []
+    source_column_block = []
+
+    # Each column should have only one source - sources listed first take
+    # precedence
+    source_columns = {}
+    for source in reversed(sources):
+        source_columns.update({col.column_name: col
+                               for col in col_data(db, source)})
+
+    qualify = (len(sources) > 1) or qualify
+    for dest_col in col_data(db, destination):
+        dest_column_block.append(INDENT + dest_col.column_name)
+        source_col = source_columns.get(dest_col.column_name)
+        if source_col:
+            if qualify:
+                source_expr = '{}.{}'.format(source_col.table_name,
+                                             source_col.column_name)
+            else:
+                source_expr = source_col.column_name
+        else:
+            source_expr = no_value(db_url)
+        if type_cast and ((not source_col) or
+                          source_col.data_type != dest_col.data_type):
+            source_expr = cast(source_expr,
+                               new_type=dest_col.data_type,
+                               db_url=db_url)
+        source_column_block.append('{}{},  -- ==> {}'.format(
+            INDENT, source_expr, dest_col.column_name))
+
+    dest_column_block = ',\n'.join(dest_column_block)
+    source_column_block = '\n'.join(source_column_block)
+    source_column_block = remove_last(source_column_block, ',')
+
+    from_clause = build_from_clause(sources)
+
+    return INSERT_TEMPLATE.format(**locals())
+
+
+VALUES_TUPLE_TEMPLATE = '''(
+{source_column_block}
+)'''
+
+INSERT_FROM_VALUES_TEMPLATE = '''
+INSERT INTO {destination} (
+{dest_column_block}
+)
+VALUES
+{source_column_blocks}'''
```

Comment:
Oh cool, I did not know `.format()` could take template strings like this--I only knew about the `"hello {}".format('world')` style. Thanks for teaching me something new!

Reply:
string templating just keeps getting better.  I hear Python 3.8 will feature entirely telepathic string templating.